### PR TITLE
i#38 attach: Enable attach on AArch64

### PR DESCRIPTION
**Pull request by derekbruening**
Implements missing functionality for ptrace attach on AArch64 and
AArch32: generated code sequences were previously x86-only;
-skip_syscall handling only supported x86; and AArch64 does not
support PTRACE_POKEUSER or PTRACE_PEEKUSER.

For AArch32, Thumb vs Arm mode require multiple steps: clearing LSB to
point at the path used as data via a call; switching to Arm mode for
DR's _start; setting the LSB of the initial app PC.

For AArch32, additionally fixes an encoder error where the opcode is
queried before copying a needs-no-encoding instruction.  This is
required for the instruction used to hold data for injection.

Tweaks the disassembler to leave a level 0 instr alone, again to
better handle the data-only insruction used for injection.

Enables the client.attach test on AArch64 and AArch32.  For AArch32,
it needs -skip_syscall.  Long-term we want that on by default
everywhere but we want explicit tests that hit it on all platforms
first.

Issue: #38

### CODE REVIEW

**Comment by derekbruening** [comment]
Will add to commit description:

Tested manually on an AArch32 machine.  Unfortunately the client.attach test is not trival to set up under QEMU with its multiple command lines and background processes so that is left as beyond the scope of this PR and part of #4719.

File: core/unix/injector.c [hunk 1]
```diff
@@ -1222,24 +1293,40 @@ injectee_run_get_retval(dr_inject_info_t *info, void *dc, instrlist_t *ilist)
...
+    if (!info->wait_syscall)
+        offset = system_call_length(app_mode);
+#    ifdef AARCH64
+    /* POKEUSER is not available. */
+    ptr_int_t saved_pc = regs.REG_PC_FIELD;
+    regs.REG_PC_FIELD = (ptr_int_t)(pc + offset);
+    r = our_ptrace_setregs(info->pid, &regs);
+    if (r < 0)
+        return r;
+    regs.REG_PC_FIELD = saved_pc;
```
Comment (bete0) [inline core/unix/injector.c:1306]: nit: If this line is not part of the if/else block, move it up before `if` or down after the `else`?

File: core/unix/injector.c [hunk 2]
```diff
@@ -1169,6 +1225,22 @@ injectee_run_get_retval(dr_inject_info_t *info, void *dc, instrlist_t *ilist)
...
+    /* Get register state before executing the shellcode. */
+    r = our_ptrace_getregs(info->pid, &regs);
+    if (r < 0)
+        return r;
+
+#    ifdef X86
+    app_mode = IF_X64_ELSE(DR_ISA_AMD64, DR_ISA_IA32);
+#    elif defined(AARCH64)
+    app_mode = DR_ISA_ARM_A64;
+#    elif defined(ARM)
```
Comment (bete0) [inline core/unix/injector.c:1239]: nit: inconsistency between using #ifdef and defined in the same block.

**Review by bete0 — APPROVED** [review]

File: core/unix/injector.c (hunk 1 above)
Comment (derekbruening) [reply core/unix/injector.c:1306]: This is part of this `if`, restoring the app's pc for the post-execution restore at the end of the function.

File: core/unix/injector.c (hunk 2 above)
Comment (derekbruening) [reply core/unix/injector.c:1239]: Fixed, here and in several other similar additions in this PR.

**Pull request merged**